[new release] alcotest-lwt, alcotest and alcotest-async (1.0.0)

CraigFe opened the pull request:
**alcotest**: Alcotest is a lightweight and colourful test framework
**alcotest-async**: Async-based helpers for Alcotest
**alcotest-lwt**: Lwt-based helpers for Alcotest

- Project page: <a href="https://github.com/mirage/alcotest/">https://github.com/mirage/alcotest/</a>
- Documentation: <a href="https://mirage.github.io/alcotest/">https://mirage.github.io/alcotest/</a>

##### CHANGES:

- Require OCaml 4.03. (mirage/alcotest#159, @hannesm)
- Change `Alcotest_{async,lwt}.test_case` to return monadic values. These must
  be run with the new `Alcotest_{async,lwt}.run` functions. See
  [examples/lwt/test.ml](https://github.com/mirage/alcotest/blob/878c500f3b25f6ebbdafc7236ebc2b7756dafe9d/examples/lwt/test.ml#L85)
  for an example of the new API. (mirage/alcotest#167, @CraigFe)
- Add generation of `latest` symlinks in the `_test` directory which point to
  the most recent test output directory. (mirage/alcotest#155, @cfcs)
- Allow all CLI options to be passed directly to `Alcotest.{run,run_with_args}`
  without use of the `argv` parameter. (mirage/alcotest#182, @CraigFe)
- Add `--compact` option for more concise result reporting. (mirage/alcotest#149,
  @andersfugmann)
- Add `--tail-errors` option for limiting the size of error logs printed to
  standard output. (mirage/alcotest#200, @mjambon)
- Add support for executing subsets of tests via the `test` subcommand. (mirage/alcotest#158,
  @CraigFe)
- Change the `float` check to include equality of `isNaN` and infinities. See
  [examples/floats.ml](https://github.com/mirage/alcotest/blob/47908b1f576d65f4418bcf778d2a1db213f5a7ff/examples/floats.ml)
  for demonstrations of the new semantics. (mirage/alcotest#152, @psafont)
- Reject test suites with colliding output directories. (mirage/alcotest#176, @CraigFe)
- Restrict the set of characters allowed in test names to alphanumerics,
  hyphens, underscores and spaces. (mirage/alcotest#161, @CraigFe)
- Fix a race condition on creating output directories. (mirage/alcotest#150, @edwintorok)
- Report test suite durations in real time rather than system time. (mirage/alcotest#162,
  @CraigFe)
- Improve documentation of the `?argv` parameter. (mirage/alcotest#164, @ian-barnes)
- Remove version number from test binary help pages. (mirage/alcotest#186, @CraigFe)
- Remove dependency on `tput`. (mirage/alcotest#189, @samoht)
- Remove result dependency. (mirage/alcotest#159, @hannesm)
- Remove uses of deprecated `Pervasives.compare`. (mirage/alcotest#173, @CraigFe)

samoht commented:
There are a few API breakage it seems -- could you check that they are expected and add an upper-bound for alcotest in the related packages?

CraigFe commented:
I believe https://github.com/ocaml/opam-repository/pull/15669/commits/537526b4ea63aceda54c29cba8fc335d8aaccf08 adds the bounds for all relevant breakages introduced here. There are a lot of spurious errors, making it difficult to know for sure.

kit-ty-kate commented on packages/alcotest-async/alcotest-async.1.0.0/opam:
```diff
@@ -0,0 +1,32 @@
+]
+
+depends: [
+  "dune"
```
```suggestion
  "dune" {>= "1.11"}
```
According to the `dune-project` file, this package requires at least dune 1.11

kit-ty-kate commented on packages/alcotest-lwt/alcotest-lwt.1.0.0/opam:
```diff
@@ -0,0 +1,31 @@
+]
+
+depends: [
+  "dune"
```
```suggestion
  "dune" {>= "1.11"}
```
According to the `dune-project` file, this package requires at least dune 1.11

kit-ty-kate commented on packages/alcotest/alcotest.1.0.0/opam:
```diff
@@ -0,0 +1,47 @@
+]
+
+depends: [
+  "dune"  {>= "1.1.0"}
```
```suggestion
  "dune" {>= "1.11"}
```
According to the `dune-project` file, this package requires at least dune 1.11

kit-ty-kate commented:
The tests seem to be a bit unstable. Could you look into it?
* https://ci.ocaml.org/log/saved/docker-run-7ac449c17e10d04de00cfbe9f299d1aa/dbe0681c5a962e21627c9d59477b22b6fff23b4c
* https://ci.ocaml.org/log/saved/docker-run-7f9f1fa22f9c824bc236e7b048b74ffe/eb4bc455d9cd094d9756f4afec006dbc9b453fe0
* https://ci.ocaml.org/log/saved/docker-run-9e98ea5459bbe1b45495ba8dd0c14f50/1c51b7898ca7f55a46127f8d64727b05bc6d697a

kit-ty-kate commented:
There are still a bunch of issues in the revdeps:
* https://ci.ocaml.org/log/saved/docker-run-c0b0a7d061dcc90111e9bae13f9a68d6/d9bf4a8612942d35e9902aedeeee3de0c4f412ad
* https://ci.ocaml.org/log/saved/docker-run-a6fb18cdaeda06aeec74be0163dfb25a/610a69a7ebe0cc8e2f3fbdffdf6c93534a585598
* https://ci.ocaml.org/log/saved/docker-run-0880a5b3283df423d645d1dd231cced3/178e7963ecbcd8d72a11a69191b15e6e02772ff6
* https://ci.ocaml.org/log/saved/docker-run-0fcf887484eb7813d9e57df35bfc38ae/4098714c0aeeb11cec97bf547b446888d61bcd68
* https://ci.ocaml.org/log/saved/docker-run-918e9fbeb0ba259bc3313d6f81ade5ae/e75867f2631c47026b70f1317cde284adccfb923
* https://ci.ocaml.org/log/saved/docker-run-7151c31d9d5c58e1174f71dda4e78306/2663e991b683c7edb060aeda3e7dee01b6590cd9
* https://ci.ocaml.org/log/saved/docker-run-cc78a10c178851db582ca9a6523aef37/e4608abeac23cdc5b84f88ad4a6a8eef53bfadd0
* https://ci.ocaml.org/log/saved/docker-run-7cf1804fc555d10ae8062724d25b5005/609c46962db741d00c740736afb2371e47a2e8c9
* https://ci.ocaml.org/log/saved/docker-run-31ca2d493cdef0b6e1bd9def51df6c75/7e51980fce53125c887603b6db77960c13154486

Some of them seem to be something else than just an API change. Could you look into them? Not also that the CI only tests packages on OCaml 4.09 so packages that requires previous versions of OCaml are not tested. If the API significantly changed it might be worth adding an upper-bound on all the packages requiring alcotest to avoid issues.

CraigFe commented:
I'm aware of the issues and am (slowly) working through them; thanks for the help triaging it all. I'll let you know when I think I've got them all.

CraigFe commented:
@kit-ty-kate: I had a look through all of the issues you listed, and am fairly sure that none of them are due to breakages in Alcotest. I added upper bounds on those to https://github.com/ocaml/opam-repository/pull/15669/commits/a3820604867ec60f08dcfae6170fc8bbab3676d5.

As you point out, without testing pre-4.09 it's difficult to know for certain which packages are broken by the (I think fairly small) API changes introduced by `alcotest-1.0.0`. As such, I added upper bounds on the rest of Alcotest's rev-deps in https://github.com/ocaml/opam-repository/pull/15669/commits/20d569c8876f4fc67eb6678e00847452b298e6c7. Let me know what you think :slightly_smiling_face:

samoht commented:
I am not sure adding upper bounds to *every* reverse dependencies is the right approach here. We want people to be able to upgrade alcotest smoothly, and this should not require the full ecosystem to be upgraded in one go.

CraigFe commented:
That's a sensible argument. I also like avoiding breaking so many different development environments in one fell swoop. 

I think it needs to be balanced with the risk of breaking large sections of the ecosystem, however; do we have a way to run pre-4.09 CI on this package's rev-deps specifically?

avsm commented:
I discussed this with @craigfe. I think it's fine to:

- merge this PR with the upper bounds
- post to discuss about the release with a list of affected packages and instructions on how to migrate, encouraging upstreams to upgrade to alcotest 1.0 
- relax upper bounds for packages which build successfully without them (i.e. try a bulk build with this PR reverted and alcotest 1.0 in the tree)

This is roughly what we do with JS Core releases, and it helps jump past the block of having to do a massive amount of work by one person to fix all the revdeps.

kit-ty-kate commented:
I opened https://github.com/ocaml/opam-repository/pull/15824 to fix the remaining revdep failures. Sorry for the time this PR took to be merged. The CI had a hard time dealing with all the new constraints so I reverted back to a smaller set of packages to constrain. More PR to fix remaining untested packages will probably be necessary but at least in most cases this should not affect users, only opam-repository CI. We can take care of that.

Thanks!
